File: sdk/apple/Sources/DesfireEV3/Authentication.swift
import CDesfire
import Foundation

/** Convert fixed C authentication metadata only after the native verifier succeeds. */
private func authenticationInfo(_ value: inout df_authentication_info_v1) throws
    -> AuthenticationInfo
{
    let transactionIdentifier = withUnsafeBytes(of: &value.transaction_identifier) { Data($0) }
    let piccCapabilities = withUnsafeBytes(of: &value.picc_capabilities) { Data($0) }
    let pcdCapabilities = withUnsafeBytes(of: &value.pcd_capabilities) { Data($0) }
    return try AuthenticationInfo(
        transactionIdentifier: transactionIdentifier,
        piccCapabilities: piccCapabilities,
        pcdCapabilities: pcdCapabilities)
}

extension Card {
    /** Establish native Standard AES from a direct, derived, or provider-resolved key. */
    public func authenticateStandardAes(
        keyNumber: KeyNumber, keySource: KeySource,
        timeoutMilliseconds: UInt32 = 5_000
    ) async throws {
        try await performWithKeySource(
            keySource, keyNumber: keyNumber, profile: .standardAes, scope: .native
        ) { handle, key in
            try withBinary(key) { pointer, count in
                var error = df_error()
                try checkNative(
                    df_authenticate_standard_aes(
                        handle, UInt32(keyNumber.rawValue), pointer, count,
                        timeoutMilliseconds, &error), error)
            }
        }
    }

    /** Establish EV2 First without PCD capability input and return verified public metadata. */
    public func authenticateEv2FirstAes(
        keyNumber: KeyNumber, keySource: KeySource,
        timeoutMilliseconds: UInt32 = 5_000
    ) async throws -> AuthenticationInfo {
        try await performWithKeySource(
            keySource, keyNumber: keyNumber, profile: .ev2First, scope: .native
        ) { handle, key in
            try withBinary(key) { pointer, count in
                var information = df_authentication_info_v1()
                information.struct_size = UInt32(MemoryLayout<df_authentication_info_v1>.size)
                information.abi_version = 1
                var error = df_error()
                try checkNative(
                    df_authenticate_ev2_first_aes(
                        handle, UInt32(keyNumber.rawValue), pointer, count,
                        timeoutMilliseconds, &information, &error), error)
                return try authenticationInfo(&information)
            }
        }
    }

    /** Establish EV2 First with zero through six explicit PCD capability bytes. */
    public func authenticateEv2FirstAesWithCapabilities(
        keyNumber: KeyNumber, keySource: KeySource,
        pcdCapabilities: Data = Data(), timeoutMilliseconds: UInt32 = 5_000
    ) async throws -> AuthenticationInfo {
        guard pcdCapabilities.count <= 6 else {
            throw DesfireError(
                .invalidArgument, "PCD capabilities must contain zero through six bytes")
        }
        return try await performWithKeySource(
            keySource, keyNumber: keyNumber, profile: .ev2First, scope: .native
        ) { handle, key in
            try withBinary(key) { keyPointer, keyCount in
                try withBinary(pcdCapabilities) { capabilityPointer, capabilityCount in
                    var information = df_authentication_info_v1()
                    information.struct_size = UInt32(
                        MemoryLayout<df_authentication_info_v1>.size)
                    information.abi_version = 1
                    var error = df_error()
                    try checkNative(
                        df_authenticate_ev2_first_aes_with_capabilities(
                            handle, UInt32(keyNumber.rawValue), keyPointer, keyCount,
                            capabilityPointer, capabilityCount, timeoutMilliseconds,
                            &information, &error), error)
                    return try authenticationInfo(&information)
                }
            }
        }
    }

    /** Replace an EV2 session through NonFirst while preserving its transaction state. */
    public func authenticateEv2NonFirstAes(
        keyNumber: KeyNumber, keySource: KeySource,
        timeoutMilliseconds: UInt32 = 5_000
    ) async throws -> AuthenticationInfo {
        try await performWithKeySource(
            keySource, keyNumber: keyNumber, profile: .ev2NonFirst, scope: .native
        ) { handle, key in
            try withBinary(key) { pointer, count in
                var information = df_authentication_info_v1()
                information.struct_size = UInt32(MemoryLayout<df_authentication_info_v1>.size)
                information.abi_version = 1
                var error = df_error()
                try checkNative(
                    df_authenticate_ev2_non_first_aes(
                        handle, UInt32(keyNumber.rawValue), pointer, count,
                        timeoutMilliseconds, &information, &error), error)
                return try authenticationInfo(&information)
            }
        }
    }

    /** Establish ISO mutual AES in PICC-master or application-key scope. */
    public func authenticateIsoAes(
        keyNumber: KeyNumber, applicationKey: Bool, keySource: KeySource,
        timeoutMilliseconds: UInt32 = 5_000
    ) async throws {
        let scope: KeyScope = applicationKey ? .isoApplication : .isoPicc
        try await performWithKeySource(
            keySource, keyNumber: keyNumber, profile: .isoAes, scope: scope
        ) { handle, key in
            try withBinary(key) { pointer, count in
                var error = df_error()
                try checkNative(
                    df_authenticate_iso_aes(
                        handle, UInt32(keyNumber.rawValue), applicationKey ? 1 : 0,
                        pointer, count, timeoutMilliseconds, &error), error)
            }
        }
    }
}

File: sdk/kotlin/src/main/kotlin/com/desfire/ev3/Authentication.kt
package com.desfire.ev3

/** Establish Standard AES with the selected native key. */
public fun BlockingCard.authenticateStandardAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
) {
    requireEmptyNativeResult(
        authenticate(AuthenticationProfile.STANDARD_AES, KeyScope.NATIVE, keyNumber, false,
            keySource, byteArrayOf(), timeoutMs),
        "Unexpected Standard AES authentication result payload",
    )
}

/** Establish EV2 First and return verified transaction and capability metadata. */
public fun BlockingCard.authenticateEv2FirstAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    pcdCapabilities: ByteArray = byteArrayOf(),
    timeoutMs: Long = 5_000,
): AuthenticationInfo {
    require(pcdCapabilities.size <= 6) { "PCD capabilities cannot exceed six bytes" }
    return AuthenticationInfo.decode(
        authenticate(AuthenticationProfile.EV2_FIRST, KeyScope.NATIVE, keyNumber, false,
            keySource, pcdCapabilities, timeoutMs),
    )
}

/** Replace an active EV2 session through NonFirst while preserving its transaction context. */
public fun BlockingCard.authenticateEv2NonFirstAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
): AuthenticationInfo = AuthenticationInfo.decode(
    authenticate(AuthenticationProfile.EV2_NON_FIRST, KeyScope.NATIVE, keyNumber, false,
        keySource, byteArrayOf(), timeoutMs),
)

/** Establish ISO mutual AES in PICC or application scope. */
public fun BlockingCard.authenticateIsoAes(
    keyNumber: KeyNumber,
    application: Boolean,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
) {
    val scope = if (application) KeyScope.ISO_APPLICATION else KeyScope.ISO_PICC
    requireEmptyNativeResult(
        authenticate(AuthenticationProfile.ISO_AES, scope, keyNumber, application,
            keySource, byteArrayOf(), timeoutMs),
        "Unexpected ISO AES authentication result payload",
    )
}

/** Invoke one profile-specific JNI entrypoint without exposing key bytes to generic dispatch. */
private fun BlockingCard.authenticate(
    profile: AuthenticationProfile,
    scope: KeyScope,
    keyNumber: KeyNumber,
    application: Boolean,
    keySource: KeySource,
    pcdCapabilities: ByteArray,
    timeoutMs: Long,
): ByteArray {
    require(timeoutMs in 1..0xFFFF_FFFFL) { "Timeout must fit a positive unsigned 32-bit value" }
    return withExclusiveHandle { handle ->
        when (keySource) {
            is KeySource.Direct -> Native.authenticateDirect(
                handle, profile.code, keyNumber.value, application, keySource.keyBytes,
                pcdCapabilities, timeoutMs,
            )
            is KeySource.Derived -> Native.authenticateDerived(
                handle, profile.code, keyNumber.value, application, keySource.masterKeyBytes,
                keySource.diversificationBytes, pcdCapabilities, timeoutMs,
            )
            is KeySource.Provider -> Native.authenticateProvider(
                handle, profile.code, scope.code, keyNumber.value, application,
                keySource.bridge, keySource.referenceBytes, keySource.diversificationBytes,
                keySource.contextBytes, keySource.applicationId?.value ?: -1,
                keySource.keySet?.value ?: -1, pcdCapabilities, timeoutMs,
            )
        }
    }
}

/** Suspend-friendly Standard AES authentication. */
public suspend fun Card.authenticateStandardAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
): Unit = call { authenticateStandardAes(keyNumber, keySource, timeoutMs) }

/** Suspend-friendly EV2 First authentication. */
public suspend fun Card.authenticateEv2FirstAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    pcdCapabilities: ByteArray = byteArrayOf(),
    timeoutMs: Long = 5_000,
): AuthenticationInfo = call {
    authenticateEv2FirstAes(keyNumber, keySource, pcdCapabilities, timeoutMs)
}

/** Suspend-friendly EV2 NonFirst authentication. */
public suspend fun Card.authenticateEv2NonFirstAes(
    keyNumber: KeyNumber,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
): AuthenticationInfo = call { authenticateEv2NonFirstAes(keyNumber, keySource, timeoutMs) }

/** Suspend-friendly ISO mutual AES authentication. */
public suspend fun Card.authenticateIsoAes(
    keyNumber: KeyNumber,
    application: Boolean,
    keySource: KeySource,
    timeoutMs: Long = 5_000,
): Unit = call { authenticateIsoAes(keyNumber, application, keySource, timeoutMs) }
